File: algorithms/same_tree.py
# Algortithm to check if two binary trees are the same

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        if p is None and q is None:
            return True
        if p is None or q is None:
            return False
        if p.val != q.val:
            return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

# Test cases
# Test case 1
p = TreeNode(1)
p.left = TreeNode(2)
p.right = TreeNode(3)
q = TreeNode(1)
q.left = TreeNode(2)
q.right = TreeNode(3)
print(Solution().isSameTree(p, q)) # True

# Test case 2 - Hard case
p = TreeNode(1)
p.left = TreeNode(2)
p.right = TreeNode(1)
q = TreeNode(1)
q.left = TreeNode(1)
q.right = TreeNode(2)
print(Solution().isSameTree(p, q)) # False

# Test case 3 - Edge case
p = None
q = None
print(Solution().isSameTree(p, q)) # True

# Test case 4 - Big tree
p = TreeNode(1)
p.left = TreeNode(2)
p.right = TreeNode(3)
p.left.left = TreeNode(4)
p.left.right = TreeNode(5)
p.right.left = TreeNode(6)
p.right.right = TreeNode(7)
q = TreeNode(1)
q.left = TreeNode(2)
q.right = TreeNode(3)
q.left.left = TreeNode(4)
q.left.right = TreeNode(5)
q.right.left = TreeNode(6)
q.right.right = TreeNode(7)
print(Solution().isSameTree(p, q)) # True

File: algorithms/cpp/same_tree.cpp
// Check if two binary trees are structurally identical with the same node values
// Time complexity:  O(n)
// Space complexity: O(h) where h is the height of the tree

#include <iostream>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v = 0, TreeNode* l = nullptr, TreeNode* r = nullptr)
        : val(v), left(l), right(r) {}
};

bool isSameTree(TreeNode* p, TreeNode* q) {
    if (p == nullptr && q == nullptr) return true;
    if (p == nullptr || q == nullptr) return false;
    if (p->val != q->val) return false;
    return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
}

int main() {
    std::cout << std::boolalpha;

    // Test case 1: same trees [1,2,3]
    TreeNode* p1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    TreeNode* q1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    std::cout << isSameTree(p1, q1) << "\n";  // true

    // Test case 2: different structure [1,2] vs [1,null,2]
    TreeNode* p2 = new TreeNode(1, new TreeNode(2), nullptr);
    TreeNode* q2 = new TreeNode(1, nullptr, new TreeNode(2));
    std::cout << isSameTree(p2, q2) << "\n";  // false

    // Test case 3: different values [1,2,3] vs [1,2,4]
    TreeNode* p3 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    TreeNode* q3 = new TreeNode(1, new TreeNode(2), new TreeNode(4));
    std::cout << isSameTree(p3, q3) << "\n";  // false
    return 0;
}
